library/java/org/chromium/net/impl/CronetBidirectionalState.java

```diff
@@ -0,0 +1,454 @@
+        // FOLLOW THROUGH
+      case Event.WRITE_COMPLETED:
+        nextAction = NextAction.INVOKE_ON_WRITE_COMPLETED_CALLBACK;
+        break;
+
+      case Event.READY_TO_FLUSH:
+        if ((originalState & State.WAITING_FOR_FLUSH) == 0) {
+          nextAction = NextAction.CARRY_ON;
+        } else {
+          nextState &= ~State.WAITING_FOR_FLUSH;
+          nextState |= State.WRITING;
+          nextAction = NextAction.SEND_DATA;
+        }
+        break;
+
+      case Event.FLUSH_DATA_COMPLETED:
+        nextState |= State.WAITING_FOR_FLUSH;
+        // FOLLOW THROUGH
+      case Event.LAST_FLUSH_DATA_COMPLETED:
+        assert (originalState & State.WRITING) != 0;
+        assert (originalState & State.WAITING_FOR_FLUSH) == 0;
+        nextState &= ~State.WRITING;
+        nextAction = NextAction.CARRY_ON;
+        break;
+
+      case Event.READY_TO_READ:
+        assert (originalState & State.WAITING_FOR_READ) == 0;
+        assert (originalState & State.READING) == 0;
+        nextState |= State.READING;
+        nextAction = (originalState & State.ON_HEADER_RECEIVED) == 0 ? NextAction.CARRY_ON
+                     : (originalState & State.END_STREAM_READ) == 0
+                         ? NextAction.READ
+                         : NextAction.INVOKE_ON_READ_COMPLETED;
+        break;
+
+      case Event.READ_COMPLETED:
+        assert (originalState & State.READING) != 0;
+        nextState &= ~State.READING;
+        nextState |= State.WAITING_FOR_READ;
+        nextAction = NextAction.INVOKE_ON_READ_COMPLETED_CALLBACK;
+        break;
+
+      case Event.LAST_READ_COMPLETED:
+        assert (originalState & State.READ_DONE) == 0;
+        nextState &= ~State.READING;
+        nextState |= State.READ_DONE;
+        nextAction = NextAction.INVOKE_ON_READ_COMPLETED_CALLBACK;
+        break;
+
+      case Event.READY_TO_FINISH:
+        if ((originalState & State.ON_COMPLETE_RECEIVED) != 0 &&
+            (originalState & State.READ_DONE) != 0 && (originalState & State.WRITE_DONE) != 0) {
+          nextState |= State.DONE;
+          nextAction = NextAction.INVOKE_ON_SUCCEEDED;
+        } else {
+          nextAction = NextAction.CARRY_ON;
+        }
+        break;
+
+      default:
+        throw new AssertionError("switch is exhaustive");
+      }
+
+      // With "Compare And Swap", the contract is the mutation succeeds only if the original value
+      // matches the expected one - this is atomic at the assembly language level: most CPUs have
+      // dedicated mnemonics for this operation - extremely efficient. And this might look like an
+      // infinite loop. It is infinite only if many Threads are eternally attempting to concurrently
+      // change the value. In fact, "Compare And Swap" is pretty bad under heavy contention - in
+      // that case it is probably better to go with "synchronized" blocks. In our case, there is
+      // none or very little contention. What matters is correctness and efficiency.
```

Comment:
This comment talks about infinite looping but it's in front of an if statement. I think if we want to keep this comment, it belongs elsewhere.

Reply:
Moved just above the `while (true)`